### PR TITLE
Setup Lerna

### PR DESCRIPTION
**Pull request by GregTheGreek**
Adds Lerna to seperate types from the beacon into their own npm module

TODO: Fix failing tests:
### Tree Structure
```
.
├── .babel-register
├── .babelrc
├── .codecov.yml
├── .dockerignore
├── .eslintrc
├── .gitignore
├── .gitmodules
├── .nycrc.json
├── .travis.yml
├── CONTRIBUTING.md
├── LICENSE
├── README.md
├── config_files
│   ├── .babel-register
│   ├── .babelrc
│   ├── .codecov.yml
│   ├── .eslintrc
│   ├── .nycrc.json
│   ├── LICENSE
│   └── tsconfig.json
├── lerna.json
├── package.json
├── packages
│   ├── eth2-types
│   │   ├── .babel-register
│   │   ├── .babelrc
│   │   ├── .codecov.yml
│   │   ├── .eslintrc
│   │   ├── .gitignore
│   │   ├── .nyc_output
│   │   ├── .nycrc.json
│   │   ├── LICENSE
│   │   ├── README.md
│   │   ├── coverage
│   │   ├── lib
│   │   ├── node_modules
│   │   ├── package.json
│   │   ├── src
│   │   ├── test
│   │   └── tsconfig.json
│   └── lodestar
│       ├── .babel-register
│       ├── .babelrc
│       ├── .codecov.yml
│       ├── .eslintrc
│       ├── .gitignore
│       ├── .nyc_output
│       ├── .nycrc.json
│       ├── LICENSE
│       ├── bin
│       ├── coverage
│       ├── docker
│       ├── keys
│       ├── node_modules
│       ├── package.json
│       ├── scripts
│       ├── src
│       ├── test
│       └── tsconfig.json
├── tsconfig.json
└── yarn.lock
```

### CODE REVIEW

File: lerna-debug.log
Comment (mpetrunic) [inline lerna-debug.log]: Lets ignore this file :smile:

File: packages/beaconChain/package.json
Comment (mpetrunic) [inline packages/beaconChain/package.json]: Let's name package directory lodestar as well?

File: lerna-debug.log
Comment (GregTheGreek) [reply lerna-debug.log]: lols wooops!

File: packages/beaconChain/package.json
Comment (GregTheGreek) [reply packages/beaconChain/package.json]: ya thats fair

File: .travis.yml
Comment (mpetrunic) [inline .travis.yml]: it will be faster if you run all lint and check-types on all packages in single command. Stage take longer to boot than executing those checks. Same goes for unit tests. e2e and spec should be run each package in it's own stage.
I would maybe suggest to run e2e and spec tests only on master/deployment

File: packages/eth2-types/package.json
Comment (mpetrunic) [inline packages/eth2-types/package.json]: license?

File: packages/eth2-types/package.json
Comment (mpetrunic) [inline packages/eth2-types/package.json]: chainsafe? haha

File: packages/eth2-types/package.json
Comment (mpetrunic) [inline packages/eth2-types/package.json]: build task is missing

File: packages/lodestar/src/chain/factory/attestation/index.ts
Comment (mpetrunic) [inline packages/lodestar/src/chain/factory/attestation/index.ts]: all this imports should be  `--- from "@chainsafe/eth2-types"`

File: packages/lodestar/package.json
Comment (mpetrunic) [inline packages/lodestar/package.json:39]: add @chainsafe/eth2-types dependency with version from package.json (it doesn't have to be published, lerna handles that)
Although if I remember correctly you have to add it using lerna

File: packages/lodestar/package.json
Comment (mpetrunic) [inline packages/lodestar/package.json:21]: Beacuse we will use symlinks for monorepo package dependencies, everything needs to be builded before executing tests

**Review by mpetrunic — COMMENTED** [review]
Some of this comments are probably premature

File: .travis.yml
Comment (GregTheGreek) [reply .travis.yml]: So here's my thoughts.... I would much rather know the specific package that fails, rather than an outright failure on like all the tests... just my two cents happy to chat it over

Comment (GregTheGreek) [reply .travis.yml]: Because as this grows into like4-5 packages, i would be much happier knowing which one failed. Also we can change this to trigger only if changes were found in specific packages later on.

File: packages/eth2-types/package.json
Comment (GregTheGreek) [reply packages/eth2-types/package.json]: boiler plate lol, ill get to this eventually

File: packages/lodestar/src/chain/factory/attestation/index.ts
Comment (GregTheGreek) [reply packages/lodestar/src/chain/factory/attestation/index.ts]: Yeah im still debuggin this stuff.... idk why its like this

File: packages/lodestar/package.json
Comment (GregTheGreek) [reply packages/lodestar/package.json:39]: So oddly emough its not working out..

File: .travis.yml
Comment (mpetrunic) [reply .travis.yml]: I wonder if we can do serialized tests (parallelization is irrelevant on this) on same machine, because with 4-5 packages we would add like 8-10 mins on CI just for booting machines to run tests

File: packages/lodestar/package.json
Comment (mpetrunic) [reply packages/lodestar/package.json:39]: lerna add @chainsafe/eth2-type --dev --scope lodestar ?

Comment (GregTheGreek) [reply packages/lodestar/package.json:39]: ,lol ya just finished this

File: packages/lodestar/package.json
Comment (GregTheGreek) [reply packages/lodestar/package.json:21]: oh interesting

Comment (mpetrunic) [reply packages/lodestar/package.json:21]: I don't remmber 100% but I think if you add build task to every package lerna can build them based on their dependency on each other

File: lerna.json
Comment (mpetrunic) [inline lerna.json]: i think if you put `independent`, you can version each package separately

File: packages/eth2-types/package.json
Comment (mpetrunic) [inline packages/eth2-types/package.json]: name this `test:unit` so we can execute unit tests in all directories with single command

File: packages/eth2-types/test/consistency.test.ts
Comment (mpetrunic) [inline packages/eth2-types/test/consistency.test.ts]: for the sake of operating systems use `path.join`

File: packages/eth2-types/test/consistency.test.ts
Comment (mpetrunic) [inline packages/eth2-types/test/consistency.test.ts]: woot? you can add describe, it to enable typing

File: packages/eth2-types/tsconfig.json
Comment (mpetrunic) [inline packages/eth2-types/tsconfig.json]: That's not needed

File: packages/lodestar/package.json
Comment (mpetrunic) [inline packages/lodestar/package.json]: why do we need this as dependency?

File: packages/lodestar/src/chain/factory/attestation/data.ts
Comment (mpetrunic) [inline packages/lodestar/src/chain/factory/attestation/data.ts]: i think all constats have wrong import. I saw this on a lot of places

**Review by mpetrunic — CHANGES_REQUESTED** [review]
remove package-lock.json in lodestar directory

File: packages/lodestar/package.json
Comment (mpetrunic) [inline packages/lodestar/package.json]: Wrong reference

File: lerna.json
Comment (GregTheGreek) [reply lerna.json]: I was thinking they should probably be together.. since updates to the client should probably be reflected on the types

File: packages/eth2-types/package.json
Comment (GregTheGreek) [reply packages/eth2-types/package.json]: oh interesting so i can do `lerna run test:unit` instead of running the scoped version?

File: packages/eth2-types/test/consistency.test.ts
Comment (GregTheGreek) [reply packages/eth2-types/test/consistency.test.ts]: could just do `import "mocha"`

File: packages/lodestar/src/chain/factory/attestation/data.ts
Comment (GregTheGreek) [reply packages/lodestar/src/chain/factory/attestation/data.ts]: oh no :(

File: packages/lodestar/package.json
Comment (GregTheGreek) [reply packages/lodestar/package.json]: we probably dont

File: lerna.json
Comment (mpetrunic) [reply lerna.json]: But then you cannot apply small fixes to libraries unless ou update client, might be irrelevant, but we could pile some fixes and then release single client update with all those fixes. Irrelevant for now

Comment (GregTheGreek) [reply lerna.json]: hmnm good point I think independent makes sense

File: packages/lodestar/coverage/lcov-report/base.css
Comment (mpetrunic) [inline packages/lodestar/coverage/lcov-report/base.css]: coverage should be ignore as well as this file

**Review by mpetrunic — CHANGES_REQUESTED** [review]

**Comment by GregTheGreek** [comment]
I've marked this `ready for review` because from a structural point of view its ready, I just have a few loose ends to close up:
1) Travis is adding a random build, labeled `nodejs/*lts`, to a few of our stages that is causing an outright failure
2) eth2-types is failing on `interface bool does not exist`
3) lodestar `test: spec` is failing because the submodules don't seem to be pulled in

**Comment by mpetrunic** [comment]
We should merge all ready pull request adn integrate here before merging though

**Comment by GregTheGreek** [comment]
> We should merge all ready pull request adn integrate here before merging though

Yeah that makes sense, all new PR's will be blocked by this one

**Pull request closed without merging**
